File: backend/src/models/JournalEntry.model.js
const mongoose = require('mongoose');

const journalEntrySchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  title: {
    type: String,
    required: [true, 'El título es requerido'],
    trim: true,
    maxlength: [200, 'El título no puede exceder 200 caracteres'],
  },
  content: {
    type: String,
    required: [true, 'El contenido es requerido'],
    maxlength: [5000, 'El contenido no puede exceder 5000 caracteres'],
  },
  mood: {
    type: String,
    enum: ['very-bad', 'bad', 'neutral', 'good', 'very-good'],
    required: [true, 'El estado de ánimo es requerido'],
  },
  aiAnalysis: {
    status: {
      type: String,
      enum: ['pending', 'processing', 'completed', 'error'],
      default: 'pending',
    },
    sentiment: {
      overall: {
        type: String,
        enum: ['negative', 'neutral', 'positive'],
      },
      score: {
        type: Number,
        min: -1,
        max: 1,
      },
      emotions: [{
        emotion: String,
        confidence: Number,
      }],
    },
    themes: [String],
    insights: String,
    processedAt: Date,
    error: String,
  },
}, {
  timestamps: true,
});

// Índices para búsquedas eficientes
journalEntrySchema.index({ userId: 1, createdAt: -1 });
journalEntrySchema.index({ 'aiAnalysis.status': 1 });

// Virtual para obtener fecha formateada
journalEntrySchema.virtual('dateFormatted').get(function () {
  return this.createdAt.toLocaleDateString('es-ES', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  });
});

// Method para marcar análisis como en proceso
journalEntrySchema.methods.startProcessing = async function () {
  this.aiAnalysis.status = 'processing';
  await this.save();
};

// Method para guardar resultado de análisis
journalEntrySchema.methods.saveAnalysis = async function (analysisResult) {
  this.aiAnalysis.status = 'completed';
  this.aiAnalysis.sentiment = analysisResult.sentiment;
  this.aiAnalysis.themes = analysisResult.themes;
  this.aiAnalysis.insights = analysisResult.insights;
  this.aiAnalysis.processedAt = new Date();
  await this.save();
};

// Method para marcar error en análisis
journalEntrySchema.methods.markAnalysisError = async function (errorMessage) {
  this.aiAnalysis.status = 'error';
  this.aiAnalysis.error = errorMessage;
  await this.save();
};

// Configurar toJSON
journalEntrySchema.set('toJSON', {
  virtuals: true,
  versionKey: false,
  transform: function (doc, ret) {
    ret.id = ret._id;
    delete ret._id;
  },
});

module.exports = mongoose.model('JournalEntry', journalEntrySchema);

File: lib/data/models/journal_entry_model.dart
class JournalEntryModel {
  final String? id;
  final String userId;
  final String title;
  final String content;
  final String mood; // 'very-bad', 'bad', 'neutral', 'good', 'very-good'
  final AIAnalysis? aiAnalysis;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  JournalEntryModel({
    this.id,
    required this.userId,
    required this.title,
    required this.content,
    required this.mood,
    this.aiAnalysis,
    this.createdAt,
    this.updatedAt,
  });

  factory JournalEntryModel.fromJson(Map<String, dynamic> json) {
    return JournalEntryModel(
      id: json['id']?.toString() ?? json['_id']?.toString(),
      userId: json['userId']?.toString() ?? '',
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      mood: json['mood'] ?? 'neutral',
      aiAnalysis: json['aiAnalysis'] != null
          ? AIAnalysis.fromJson(json['aiAnalysis'])
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'userId': userId,
      'title': title,
      'content': content,
      'mood': mood,
      if (aiAnalysis != null) 'aiAnalysis': aiAnalysis!.toJson(),
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }
}

class AIAnalysis {
  final String status; // 'pending', 'processing', 'completed', 'error'
  final Sentiment? sentiment;
  final List<String> themes;
  final String? insights;
  final DateTime? processedAt;
  final String? error;

  AIAnalysis({
    required this.status,
    this.sentiment,
    this.themes = const [],
    this.insights,
    this.processedAt,
    this.error,
  });

  factory AIAnalysis.fromJson(Map<String, dynamic> json) {
    return AIAnalysis(
      status: json['status'] ?? 'pending',
      sentiment: json['sentiment'] != null
          ? Sentiment.fromJson(json['sentiment'])
          : null,
      themes: (json['themes'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      insights: json['insights'],
      processedAt: json['processedAt'] != null
          ? DateTime.parse(json['processedAt'])
          : null,
      error: json['error'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      if (sentiment != null) 'sentiment': sentiment!.toJson(),
      'themes': themes,
      if (insights != null) 'insights': insights,
      if (processedAt != null) 'processedAt': processedAt!.toIso8601String(),
      if (error != null) 'error': error,
    };
  }
}

class Sentiment {
  final String overall; // 'negative', 'neutral', 'positive'
  final double score; // -1 to 1

  Sentiment({
    required this.overall,
    required this.score,
  });

  factory Sentiment.fromJson(Map<String, dynamic> json) {
    return Sentiment(
      overall: json['overall'] ?? 'neutral',
      score: (json['score'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'overall': overall,
      'score': score,
    };
  }
}

class JournalStatsModel {
  final int totalEntries;
  final int entriesThisMonth;
  final bool canCreateMore;
  final dynamic remaining; // int or 'unlimited'
  final bool isPremium;
  final List<MoodDistribution> moodDistribution;
  final List<ThemeCount> topThemes;

  JournalStatsModel({
    required this.totalEntries,
    required this.entriesThisMonth,
    required this.canCreateMore,
    this.remaining,
    required this.isPremium,
    this.moodDistribution = const [],
    this.topThemes = const [],
  });

  factory JournalStatsModel.fromJson(Map<String, dynamic> json) {
    final stats = json['stats'] as Map<String, dynamic>? ?? {};

    return JournalStatsModel(
      totalEntries: stats['totalEntries'] ?? 0,
      entriesThisMonth: stats['entriesThisMonth'] ?? 0,
      canCreateMore: stats['canCreateMore'] ?? false,
      remaining: stats['remaining'],
      isPremium: stats['isPremium'] ?? false,
      moodDistribution: (json['moodDistribution'] as List<dynamic>?)
              ?.map((e) => MoodDistribution.fromJson(e))
              .toList() ??
          [],
      topThemes: (json['topThemes'] as List<dynamic>?)
              ?.map((e) => ThemeCount.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class MoodDistribution {
  final String mood;
  final int count;

  MoodDistribution({
    required this.mood,
    required this.count,
  });

  factory MoodDistribution.fromJson(Map<String, dynamic> json) {
    return MoodDistribution(
      mood: json['_id'] ?? json['mood'] ?? '',
      count: json['count'] ?? 0,
    );
  }
}

class ThemeCount {
  final String theme;
  final int count;

  ThemeCount({
    required this.theme,
    required this.count,
  });

  factory ThemeCount.fromJson(Map<String, dynamic> json) {
    return ThemeCount(
      theme: json['theme'] ?? '',
      count: json['count'] ?? 0,
    );
  }
}
